File: JANUARY/09-01-2026/865.rb
def subtree_with_all_deepest(root)
  dfs = lambda do |node|
    return [0, nil] if node.nil?
    ld, ln = dfs.call(node.left)
    rd, rn = dfs.call(node.right)
    return [ld + 1, ln] if ld > rd
    return [rd + 1, rn] if rd > ld
    [ld + 1, node]
  end

  dfs.call(root)[1]
end

File: JANUARY/09-01-2026/865.kt
class Solution {
    fun subtreeWithAllDeepest(root: TreeNode?): TreeNode? {
        fun dfs(node: TreeNode?): Pair<Int, TreeNode?> {
            if (node == null) return Pair(0, null)
            val (ld, ln) = dfs(node.left)
            val (rd, rn) = dfs(node.right)
            return when {
                ld > rd -> Pair(ld + 1, ln)
                rd > ld -> Pair(rd + 1, rn)
                else -> Pair(ld + 1, node)
            }
        }
        return dfs(root).second
    }
}